staging/src/k8s.io/apiserver/pkg/admission/plugin/cel/controller.go

```diff
@@ -0,0 +1,417 @@
+/*
+Copyright 2022 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package cel
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"sync"
+	"time"
+
+	corev1 "k8s.io/api/core/v1"
+	k8serrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/api/meta"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	utilruntime "k8s.io/apimachinery/pkg/util/runtime"
+	"k8s.io/apimachinery/pkg/util/sets"
+	"k8s.io/apiserver/pkg/admission"
+	"k8s.io/apiserver/pkg/admission/plugin/cel/generic"
+	"k8s.io/client-go/dynamic"
+	"k8s.io/client-go/dynamic/dynamicinformer"
+	"k8s.io/client-go/tools/cache"
+)
+
+// celAdmissionController is the top-level controller for admission control using CEL
+// it is responsible for watching policy definitions and bindings, and for each
+// binding, watches
+type celAdmissionController struct {
+	// Context under which the controller runs
+	runningContext context.Context
+
+	policyDefinitionsController generic.Controller[fantasyInternalPolicyDefinition]
+	policyBindingController     generic.Controller[fantasyInternalPolicyBinding]
+	paramsCRDControllers        map[schema.GroupVersionKind]paramInfo
+
+	restMapper meta.RESTMapper
+
+	// dynamicclient used to create informers to watch the param crd types
+	dynamicClient  dynamic.Interface
+	openapiFetcher func(schema.GroupVersionResource) interface{}
+
+	// Index for each definition namespace/name, contains all binding
+	// namespace/names known to exist for that definition
+	definitionInfo map[string]definitionInfo
+
+	bindingInfos map[string]bindingInfo
+
+	mutex sync.Mutex
+}
+
+type definitionInfo struct {
+	paramGr           schema.GroupVersionKind
+	dependentBindings sets.String
+}
+
+type bindingInfo struct {
+	definitionName string
+
+	// Compiled CEL expressoin turned into an evaluator
+	evaluator EvaluatorFunc
+}
+
+type paramInfo struct {
+	controller           generic.Controller[*unstructured.Unstructured]
+	stop                 func()
+	dependentDefinitions sets.String
+}
+
+func NewAdmissionController(
+	//
+	policyDefinitionsInformer cache.SharedIndexInformer,
+	policyBindingInformer cache.SharedIndexInformer,
+
+	// Fetches OpenAPI specifications for a given GroupVersionResource
+	//
+	openapiFetcher func(schema.GroupVersionResource) interface{},
```

Comment:
Admission Controller needs some way to provide the `Compile` function with open api schemas (first for params, in the future also for the object types themselves: both crd and native)

some questions I have about this:

1.  what type we choose to use to represent the openapi? (could choose gnostic for example or kube-openapi types? structural schema? is smdschema an option?)
2. how to keep schemas for non-native types up-to-date? we won't be able to watch CRDs in this package.

Reply:
> what type we choose to use to represent the openapi? (could choose gnostic for example or kube-openapi types? structural schema? is smdschema an option?)

The `Compile` implementation will ultimately want CEL `Decl` types, is currently able to convert `Structural` to them, and could be taught to convert [extensions.v1.JsonSchema](https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1/types_jsonschema.go).  Maybe this is an example of where we want to do some sort of `DeclConverter` interface that apiextensions can inject into the apiserver and that can be then used to do something like `Compile(injectedDeclConverter(paramsSchema))`?

>how to keep schemas for non-native types up-to-date? we won't be able to watch CRDs in this package (presumably that logic can be punted to whatever implements the fetcher callback)

Putting the logic in the apiextensions server and then injecting it in here via some interface (fetcher callback or whatever) sounds like the only option to me.

If a basic apiserver is created that does not have an apiextensions server, there will be no params support available. It would be possible to have param-less policies. It's a good case to add a test for (at some point).

Reply:
```
Putting the logic in the apiextensions server and then injecting it in here via some interface 
```
Will this cause dependency loop?

Reply:
I dont think so. The place it would need to be injected is inside `kubeapiserver`, which sits outside of the dependency loop.  The injection would need to be placed somewhere in here:

https://github.com/kubernetes/kubernetes/blob/47ff6c6590932930f47b24e6c3769b98d0aa4d97/pkg/kubeapiserver/admission/config.go#L50-L82